src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/UseNameofInPlaceOfString.cs

```diff
@@ -0,0 +1,152 @@
+                    if (HasAMatchInScope(stringText, parametersInScope))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(
+                            RuleWithSuggestion, argument.Value.Syntax.GetLocation(), properties, stringText ));
+                    }
+                    return;
+                case PropertyName:
+                    var propertiesInScope = GetPropertiesInScope(context);
+                    if (HasAMatchInScope(stringText, propertiesInScope))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(
+                            RuleWithSuggestion, argument.Value.Syntax.GetLocation(), properties, stringText));
+                    }
+                    return;
+                default:
+                    return;
+            }
+        }
+
+        private IEnumerable<string> GetPropertiesInScope(OperationAnalysisContext context)
+        {
+            var containingType = context.ContainingSymbol.ContainingType;
+            // look for all of the properties in the containing type and return the property names
+            if (containingType != null)
+            {
+                foreach (var property in containingType.GetMembers().OfType<IPropertySymbol>())
+                {
+                    yield return property.Name;
+                }
+            }
+        }
+
+        internal IEnumerable<string> GetParametersInScope(OperationAnalysisContext context)
+        {
+            // get the parameters for the containing method
+            foreach (var parameter in context.ContainingSymbol.GetParameters())
+            {
+                yield return parameter.Name;
+            }
+
+            // and loop through the ancestors to find parameters of anonymous functions and local functions
+            var parentOperation = context.Operation.Parent;
+            while (parentOperation != null)
+            {
+                if (parentOperation.Kind == OperationKind.AnonymousFunction)
+                {
+                    var lambdaSymbol = ((IAnonymousFunctionOperation)parentOperation).Symbol;
+                    if (lambdaSymbol != null)
+                    {
+                        foreach (var lambdaParameter in lambdaSymbol.Parameters)
+                        {
+                            yield return lambdaParameter.Name;
+                        }
+                    }
+                }
+                else if (parentOperation.Kind == OperationKind.LocalFunction)
+                {
+                    var localFunction = ((ILocalFunctionOperation)parentOperation).Symbol;
+                    foreach (var localFunctionParameter in localFunction.Parameters)
+                    {
+                        yield return localFunctionParameter.Name;
+                    }
+                }
+
+                parentOperation = parentOperation.Parent;
```

Comment:
should this loop ends when it meets member decl? and this seems done first before douing context.ContainingSymbol.GetParameters?

Reply:
It won't reach member decl because it is IOperation and only works on executable code.  Also, the order of the parameters doesn't matter as long as we get all of them.

Reply:
oops. sorry, I thought it was syntax node :)

Reply:
No problem.  Thanks for the review!